pennylane/operation.py

```diff
@@ -300,94 +439,316 @@ def hash(self):
         """int: returns an integer hash uniquely representing the operator"""
         return hash((str(self.name), tuple(self.wires.tolist()), _process_data(self)))
 
-    @classmethod
-    def _matrix(cls, *params):
-        """Matrix representation of the operator
-        in the computational basis.
+    @staticmethod
+    def compute_matrix(*params, **hyperparams):  # pylint:disable=unused-argument
+        """Canonical matrix of this operator in the computational basis.
+
+        The canonical matrix is the textbook matrix representation that does not consider wires.
+        Implicitly, this assumes that the wires of the operator correspond to the global wire order.
+
+        .. note::
+            This method gets overwritten by subclasses to define the matrix representation
+            of a particular operator.
 
-        This is a *class method* that should be defined for all
-        new operations and observables, that returns the matrix representing
-        the operator in the computational basis.
+        Args:
+            params (list): trainable parameters of this operator, as stored in ``op.parameters``
+            hyperparams (dict): non-trainable hyperparameters of this operator, as stored in ``op.hyperparameters``
 
-        This private method allows matrices to be computed
-        directly without instantiating the operators first.
+        Returns:
+            tensor_like: matrix representation
 
-        To return the matrices of *instantiated* operators,
-        please use the :attr:`~.Operator.matrix` property instead.
+        **Example**
 
-        **Example:**
+        >>> qml.CNOT.compute_matrix()
+        [[1, 0, 0, 0],
+        [0, 1, 0, 0],
+        [0, 0, 0, 1],
+        [0, 0, 1, 0]]
 
-        >>> qml.RY._matrix(0.5)
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        The matrix representation may depend on parameters or hyperparameters:
+
+        >>> qml.Rot.compute_matrix(0.1, 0.2, 0.3)
+        [[ 0.97517033-0.19767681j -0.09933467+0.00996671j]
+         [ 0.09933467+0.00996671j  0.97517033+0.19767681j]]
+
+        If parameters are tensors, a tensor of the same type is returned:
+
+        >>> res = qml.Rot.compute_matrix(torch.tensor(0.1), torch.tensor(0.2), torch.tensor(0.3))
+        >>> type(res)
+        <class 'torch.Tensor'>
+        """
+        raise MatrixUndefinedError
+
+    def matrix(self, wire_order=None):
```

Comment:
I think this is great new functionality. I just beg you to make this change such that it is possible to write custom Operations and devices in a way that is compatible with both the old an new Operation class. There is a huge number of PL devices out there (and this is awesome!), it would be very sad if all those devices stopped working after this refactor or could only be implemented to work with either the new or the old version of Operation. Please, please, please be mindful of the community here and test this extra carefully with a large number of third party devices.

Reply:
It is just a matter of naming things right. For example: Maybe this method could be called `compute_matrix()` and  `.matrix` could remain a property of Operation that simply returns  `self.compute_matrix(wire_ordering=None)`. Similar logic could be applied elsewhere.

Reply:
@cvjjm I'll move the discussion over to the main branch soon so it does not get lost.